### PR TITLE
prov/efa: hmem performance enhancement when using gdrcopy

### PR DESCRIPTION
**Pull request by wzamazon**
This patch include 3 commits that aimed to enhance hmem performance when gdrcopy is used.

### CODE REVIEW

File: prov/efa/src/rxr/rxr.h [hunk 1]
```diff
@@ -538,6 +539,22 @@ struct rxr_domain {
...
+ * copies, and improve performance.
+ */
+struct rxr_queued_copy {
+	struct rxr_pkt_entry *pkt_entry;
+	char *data;
+	size_t data_size;
+	size_t data_offset;
+};
+
+#define RXR_EP_MAX_QUEUED_COPY (8)
```
Comment (shijin-aws) [inline prov/efa/src/rxr/rxr.h:556]: Why 8 is the chosen value here?

File: prov/efa/src/rxr/rxr_pkt_type_base.c [hunk 2]
```diff
@@ -260,8 +267,28 @@ ssize_t rxr_pkt_copy_data_to_rx_entry(struct rxr_ep *ep,
...
+	 *    available to copy. In the case, this function is still called because
+	 *    receiving end point did not notify the sender about the cancelation,
+	 *    thus sender will keep sending.
+	 *
+	 * 2. application's receiving buffer is smaller than incoming message size,
+	 *    and this function will not copy to beyond receiving buffer.
+	 *    Again, in the case, receiving end point did not notify the sender
+	 *    about the truncation, thus sender will keep sending.
+	 *
+	 * 3. message size is 0, thus not data to copy
```
Comment (shijin-aws) [inline prov/efa/src/rxr/rxr_pkt_type_base.c:372]: `not` -> `no`

File: prov/efa/src/rxr/rxr_pkt_type_base.c [hunk 3]
```diff
@@ -260,8 +267,28 @@ ssize_t rxr_pkt_copy_data_to_rx_entry(struct rxr_ep *ep,
...
-	if (data_size == 0)
-		goto out;
+	/*
+	 * Under 3 rare situations, this function does not perform the copy
+	 * action, but still consider data is copied:
+	 *
+	 * 1. application cancelled the receive, thus the receive buffer is not
+	 *    available to copy. In the case, this function is still called because
+	 *    receiving end point did not notify the sender about the cancelation,
+	 *    thus sender will keep sending.
```
Comment (shijin-aws) [inline prov/efa/src/rxr/rxr_pkt_type_base.c:365]: I would rephrase these to  
> this function is still called, otherwise the sender will keep sending as the receiving end point did not notify the sender about the cancelation

Similar to the scenario 2.

File: prov/efa/src/rxr/rxr_pkt_type_base.c [hunk 4]
```diff
@@ -260,8 +267,28 @@ ssize_t rxr_pkt_copy_data_to_rx_entry(struct rxr_ep *ep,
...
+	 *    and this function will not copy to beyond receiving buffer.
+	 *    Again, in the case, receiving end point did not notify the sender
+	 *    about the truncation, thus sender will keep sending.
+	 *
+	 * 3. message size is 0, thus not data to copy
+	 */
+	if (OFI_UNLIKELY((rx_entry->rxr_flags & RXR_RECV_CANCEL)) ||
+	    OFI_UNLIKELY(data_offset >= rx_entry->cq_entry.len) ||
+	    OFI_UNLIKELY(data_size == 0)) {
+		rxr_pkt_handle_data_copied(ep, pkt_entry, data_size);
```
Comment (shijin-aws) [inline prov/efa/src/rxr/rxr_pkt_type_base.c:377]: Why not keep the `out:` section and `goto out`?

File: prov/efa/src/rxr/rxr_pkt_type_base.c [hunk 5]
```diff
@@ -229,19 +229,107 @@ size_t rxr_pkt_data_size(struct rxr_pkt_entry *pkt_entry)
...
+ *    will post a local read request to copy data. (This is because NCCL forbids its
+ *    plugin to make cuda calls). In this case, the data is not copied upon return of
+ *    this function, and the function rxr_pkt_handle_copied() is not called. It will
+ *    be called upon the completion of local read operation by the progress engine.
+ *
+ *    Otherwise, this function calls rxr_pkt_copy_data_to_hmem(), which will batch
+ *    multiple copies, and perform the copy (then call rxr_pkt_handle_copied()) together
+ *    to improve performance.
+ *
+ * When application's receive buffer is on device, data is copied immediately, and
```
Comment (shijin-aws) [inline prov/efa/src/rxr/rxr_pkt_type_base.c:331]: I am a bit confused whether the L331-332 is a duplicate statement which has been covered by the two scenarios above already? Are you saying `When application's receive buffer is NOT on device, data is copied immediately,`?

File: prov/efa/src/rxr/rxr.h (hunk 1 above)
Comment (wzamazon) [reply prov/efa/src/rxr/rxr.h:556]: Experiments show 8 yielded best performance.

File: prov/efa/src/rxr/rxr_pkt_type_base.c (hunk 4 above)
Comment (wzamazon) [reply prov/efa/src/rxr/rxr_pkt_type_base.c:377]: I generally try to avoid using goto.

File: prov/efa/src/rxr/rxr_pkt_type_base.c (hunk 3 above)
Comment (wzamazon) [reply prov/efa/src/rxr/rxr_pkt_type_base.c]: Sure,  but I would change "otherwise" to "because".

File: prov/efa/src/rxr/rxr_pkt_type_base.c (hunk 5 above)
Comment (wzamazon) [reply prov/efa/src/rxr/rxr_pkt_type_base.c]: Yes, you are right. Fixed

File: prov/efa/src/rxr/rxr_pkt_type_base.c (hunk 2 above)
Comment (wzamazon) [reply prov/efa/src/rxr/rxr_pkt_type_base.c]: Fixed in new revision.

File: prov/efa/src/rxr/rxr_pkt_type_base.c (hunk 3 above)
Comment (shijin-aws) [reply prov/efa/src/rxr/rxr_pkt_type_base.c]: In that case let's say
> because the sender will keep sending if the receiving end point did not notify the sender about the cancelation

Comment (wzamazon) [reply prov/efa/src/rxr/rxr_pkt_type_base.c]: It cannot be "if" because receiver for sure did not notify the sender about the cancellation.

Comment (shijin-aws) [reply prov/efa/src/rxr/rxr_pkt_type_base.c]: I see. Sounds good.

File: prov/efa/src/rxr/rxr.h (hunk 1 above)
Comment (shijin-aws) [reply prov/efa/src/rxr/rxr.h:556]: OK, thanks.
